README.md
# methtuple

![Python package](https://github.com/PeteHaitch/methtuple/actions/workflows/python.yaml/badge.svg)

## Overview

### What does it do?

`methtuple` allows the user to investigate the co-occurence of methylation marks at the level of individual DNA fragments. It does this by performing methylation calling at _m-tuples_ of methylation loci from high-throughput bisulfite sequencing data, such as _methylC-seq_. In short, `methtuple` extracts and tabulates the methylation states of all m-tuples from a SAM/BAM file (for a user-defined value of _m_).

### Why would I want to do that?

A typical read from a bisulfite-sequencing experiment reports a binary methylated or unmethylated measurement at multiple loci. Each read originates from a single cell. Because methylation calls are made from individual reads/read-pairs, we can investigate the co-occurence of methylation events at the level of individual DNA fragments.

I have been using `methtuple` to investigate the spatial dependence of DNA methylation at the level of individual DNA fragments by studying methylation patterns of CpG 2-tuples. `methtuple` can also be used as a drop-in replacement for `bismark_methylation_extractor` while also providing enhanced filtering options and a slightly faster runtime (10-20% faster, albeit with an increased memory usage).

### What is an m-tuple?

The simplest _m-tuple_ is the 1-tuple (_m_ = 1). `methtuple` tabulates the number of reads that are methylated (_M_) and unmethylated (_U_) for each methylation 1-tuple in the genome. 1-tuples are the type of methylation calling performed by most methylation calling software such as Bismark's `bismark_methylation_extractor`.

A 2-tuple (_m_ = 2) is a pair of methylation loci. `methtuple` tabulates the number of reads that methylated at each locus in the pair (_MM_), both unmethylated (_UU_) or methylated at one locus but not the other (_MU_ or _UM_). This idea readily extends to 3-tuples, 4-tuples, etc.

In its default settings, and with _m_ > 1, `methtuple` tries to create only m-tuples made of "neighbouring" loci. However, please see the example below for why I say this only "tries" to create m-tuples of neighbouring loci. For a DNA fragment containing _k_ methylation loci there are _k - m + 1_ m-tuples made of neighbouring loci.

Alternatively, we can create all combinations of m-tuples by using the `--all-combinations` flag. For a DNA fragment containing _k_ methylation loci there are "_k_ choose _m_" m-tuples when using `--all-combinations`, a number that grows rapidly in _k_, particularly when _m_ is close to _k/2_.

Regardless of how m-tuples are constructed, `methtuple` always takes care to only count each methylation locus once when it has been twice-sequenced by overlapping paired-end reads.

### Draw me a picture

Well, I hope ASCII art will do.

Suppose we sequence a region of the genome containing five methylation loci with three paired-end reads (`A`, `B` and `C`):

```text
ref: 1    2   3 4 5
A_1: |----->
A_2:         <------|
B_1: |----->
B_2:           <----|
C_1:    |----->
C_2:      <------|
```

If we are interested in 1-tuples, then we would obtain the following from each read by running `methtuple`:

```text
A: {1}, {2}, {3}, {4}, {5}
B: {1}, {2}, {4}, {5}
C: {2}, {3}, {4}
```

This result is true regardless of whether the `--all-combinations` flag is set.

If we are interested in 3-tuples, then we would obtain the following from each read by running `methtuple` in its default mode:

```text
A: {1, 2, 3}, {2, 3, 4}, {3, 4, 5}
B: {1, 2, 4}, {2, 4, 5}
C: {2, 3, 4}
```

Things to note:

* Read-pair `A` sequences all three (= 5 - 3 + 1) "neighbouring" 3-tuples
* Read-pair `B` sequences none of the "neighbouring" 3-tuples but does "erroneously" construct two non-neighbouring 3-tuples. This happens because m-tuples are created independently from each read-pair; effectively, read-pair `B` is "unaware" of methylation locus `3`. Depending on the downstream analysis, you may want to _post-hoc_ filter out these "non-neighbouring" m-tuples.
* The twice-sequenced methylation loci, `2` and `3`, in read-pair `C` are not double counted.

However, if we were to run `methtuple` with `--all-combinations` then we would obtain:

```text
A: {1, 2, 3}, {2, 3, 4}, {3, 4, 5}, {1, 2, 4}, {1, 2, 5}, {1, 3, 4}, {1, 3, 5}, {1, 4, 5}, {2, 3, 5}, {2, 4, 5}
B: {1, 2, 4}, {2, 4, 5}, {1, 2, 5}, {1, 4, 5}
C: {2, 3, 4}
```

## Installation and dependencies

`methtuple` is written in Python and relies upon the `pysam` module. __NOTE: `methtuple` now requires `pysam v0.8.4` or greater.__

Running `python setup.py install` will attempt to install `pysam` if it isn't found on your system. Alternatively, instructions for installing `pysam` are available from [https://github.com/pysam-developers/pysam](https://github.com/pysam-developers/pysam).

I have extensively used and tested `methtuple` (<= `v.1.6.0) with Python 2.7 and it should also work on Python 3.x.
As of `v1.7.0` I stopped trying to provide compatability with Python 2.x.
Although, I have not used the software much since 2015, it passes its extensive test suite https://github.com/PeteHaitch/methtuple/actions and I will endeavour to provide fixes to any bug reports.

### Using `pip`

The simplest way:

```sh
pip install methtuple
```

`methtuple` is written in Python and requires the `pysam` module. __NOTE: `methtuple` now requires `pysam v0.8.4` or greater.__

Alternatively, after cloning or downloading the `methtuple` git repositority, simply run:

```sh
python setup.py install
```

in the root `methtuple` directory should work for most systems.

## Usage

### Basic usage

`methtuple` processes a single SAM or BAM file and works for both single-end and paired-end sequencing data. Example BAM files from single-end directional and paired-end directional bisulfite-sequencing experiments are available in the `data/` directory.

Methylation measurements may be filtered by base quality or other criteria such as the mapping quality of the read or whether the read is marked as a PCR duplicate. For a full list of filtering options, please run `methtuple --help` or see the __Advanced Usage__ section below.

Currently, the SAM/BAM file must have been created with [Bismark](http://www.bioinformatics.bbsrc.ac.uk/projects/download.html#bismark). If the data were aligned with Bismark version < 0.8.3 please use the `--aligner Bismark_old` flag. Please file an issue if you would like to use a SAM/BAM file created with another aligner and I will do my best to support it.

The main options to pass `methtuple` are the size of the m-tuple (`-m`); the type of methylation, which is some combination of _CG_, _CHG_, _CHH_ and _CNN_ (`--methylation-type`); any filters to be applied to reads or positions within reads (see below); the SAM/BAM file; and the sample name, which will be used as a prefix for all output files. Multiple methylation types may be specified jointly, e.g., `--methylation-type CG --methylation-type CHG`

### Output

Three output files are created and summary information is written to `STDOUT`. The main output file is a tab-delimited file of all m-tuples, `<in>.<--methylation-type>.<-m>[ac].tsv`, where `<in>` is the prefix of the `<in.bam>|<in.sam>` SAM/BAM file and `ac` is added if the `--all-combinations` flag was used, e.g., `SRR949207.CG.2ac.tsv`. Output files may be gzipped (`--gzip`) or bzipped (`--bzip2`).

Here are the first 5 rows (including with the header row) from `data/se_directional.fq.gz_bismark_bt2.CG.2.tsv`, which is created by running the single-end directional example shown below:

```text
chr     strand  pos1    pos2    MM      MU      UM      UU
chr1    +       6387768 6387783 1       0       0       0
chr1    +       7104116 7104139 1       0       0       0
chr1    +       7104139 7104152 1       0       0       0
chr1    +       9256170 9256179 0       0       0       1
```

So, for example, at the CpG 2-tuple chr1:+:(6,387,768, 6,387,783) we observed 1 read that was methylated at chr1:+:6,387,768 and methylated at chr1:+:6,387,783.

The `strand` is recorded as `+` (forward strand, "OT" in Bismark), `-` (reverse strand, "OB" in Bismark) or `*`, meaning not applicable (if the `--strand-collapse` option is set). The position of all methylation loci is always with respect to the forward strand.

The second file (`<in>.<--methylation-type>_per_read.hist`) is a text histogram of the number of methylation loci per read/readpair (of the type specified by `--methylation-type`) that passed the filters specified at runtime of `methtuple`.

Here is the file `data/se_directional.fq.gz_bismark_bt2.CG_per_read.hist`, which is created by running the single-end directional example shown below:

```text
n       count
0       4561
1       2347
2       789
3       296
4       144
5       61
6       29
7       19
8       3
9       4
10      2
11      1
12      3
13      4
14      1
18      2
```

So, 4,561 reads aligned to a position containing no CpGs while 2 reads aligned to a position containing 18 CpGs.

An optional third and final file (`<in>.reads_that_failed_QC.txt>`) records the querynames (`QNAME`) of all reads that failed to pass quality control filters and which filter the read failed. This file may be omitted by use of the `--no-failed-filter-file` flag.

In this case we didn't set any quality control filters and so this file is empty.

### Examples

Two small example datasets are included in the `data/` directory. Included are the FASTQ files and the SAM/BAM files generated with __Bismark__ in __Bowtie2__ mode. More details of the example datasets can be found in `data/README.md`

Although the example datasets are both from directional bisulfite-sequencing protocols, `methtuple` also works with data from non-directional bisulfite-sequencing protocols.

#### Single-end reads

The following command will extract all CpG 2-tuples from the file `data/se_directional.bam`:

```sh
methtuple -m 2 --methylation-type CG data/se_directional.fq.gz_bismark_bt2.bam
```

This results in 3 files:

* `data/se_directional.fq.gz_bismark_bt2.CG.2.tsv`
* `data/se_directional.fq.gz_bismark_bt2.CG_per_read.hist`
* `data/se_directional.fq.gz_bismark_bt2.reads_that_failed_QC.txt`

#### Paired-end reads

Paired-end data must firstly be sorted by queryname prior to running `methtuple`. BAM files created by Bismark, such as `data/pe_directional.bam`, are already sorted by queryname. So, to extract all CG/CHH 3-tuples we would simply run:

```sh
methtuple -m 3 --methylation-type CG --methylation-type CHH data/pe_directional_1.fq.gz_bismark_bt2_pe.bam
```

This results in 3 files:

* `data/pe_directional_1.fq.gz_bismark_bt2_pe.CG_CHH.3.tsv`
* `data/pe_directional_1.fq.gz_bismark_bt2_pe.CG_CHH_per_read.hist`
* `data/pe_directional_1.fq.gz_bismark_bt2_pe.reads_that_failed_QC.txt`

##### Note on sort-order of paired-end SAM/BAM files

If your paired-end SAM/BAM file is sorted by genomic coordinates, then you must first sort the SAM/BAM by queryname and then run `methtuple` on the queryname-sorted SAM/BAM. This can be done by using `samtools sort` with the `-n` option or Picard's `SortSam` function with the `SO=queryname` option:

```sh
# Create a coordinate-sorted SAM/BAM for the sake of argument
samtools sort data/pe_directional_1.fq.gz_bismark_bt2_pe.bam data/cs_pe_directional_1.fq.gz_bismark_bt2_pe
# Re-sort the coordinate-sorted BAM by queryname
samtools sort -n data/cs_pe_directional_1.fq.gz_bismark_bt2_pe.bam data/qs_pe_directional_1.fq.gz_bismark_bt2_pe
# Run methtuple on the queryname sorted BAM
methtuple -m 3 --methylation-type CG --methylation-type CHG data/qs_pe_directional_1.fq.gz_bismark_bt2_pe.bam
```

### Memory usage and running time

For a rough indication of performance, here are the results for processing approximately 40,000,000 100bp paired-end reads from chr1 of a 20-30x coverage whole-genome methylC-seq experiment of human data. This analysis used a single AMD Opteron 6276 CPU (2.3GHz) on a shared memory system.

#### `-m 2`

Memory usage peaked at 1.9GB and the running time was approximately 5 hours.

#### `-m 2 --all-combinations`

Memory usage peaked at 7GB and the running time was approximately 5.5 hours.

Use of the `--all-combinations` flag creates all possible m-tuples, including non-neighbouring ones. This produces many more m-tuples and so increases the memory usage.

#### `-m 5`

Memory usage peaked at 1.5GB and the running time was approximately 4.3 hours.

### Helper script

I frequently work with large, coordinate-sorted SAM/BAM files. To speed up the extraction of m-tuples, I use a simple parallelisation strategy with [GNU parallel](http://www.gnu.org/software/parallel/). The idea is to split the SAM/BAM file into chromosome-level SAM/BAM files, process each chromosome-level SAM/BAM separately and then recombine these chromosome-level files into a genome-level file. The script `helper_scripts/run_methtuple.sh` implements this strategy; simply edit the key variables in this script or adapt it to your own needs. Please check the requirements listed in `helper_scripts/run_methtuple.sh`.

#### Warnings

* __WARNING__: This simple strategy uses as many cores as there are chromosomes. This can result in __very__ large memory usage, depending on the value of `-m`, and may cause problems if you have more chromosomes than available cores.
* __WARNING__: The script `tabulate_hist.R` must be in the same directory as `run_methtuple.sh`

### Advanced usage

A full list of options is available by running `methtuple --help`:

```text
usage: methtuple [options] <in.bam>|<in.sam>
Please run 'methtuple -h' for a full list of options.

Extract methylation patterns at m-tuples of methylation loci from the aligned
reads of a bisulfite-sequencing experiment. Currently only supports SAM/BAM
files created with Bismark.

Input options:
  <in.bam>|<in.sam>     Input file in BAM or SAM format. Use - to specify STDIN.
                        The header must be included and alignments must have
                        been done using Bismark.
  --aligner {Bismark,Bismark_old}
                        The aligner used to generate the SAM/BAM file.
                        Bismark_old refers to Bismark version < 0.8.3 (default:
                        Bismark)
  --Phred64             Quality scores are encoded as Phred64 rather than
                        Phred33 (default: False)

Output options:
  -o PREFIX, --output-prefix PREFIX
                        By default, all output files have the same prefix as
                        that of the input file. This will override the prefix of
                        output file names
  --sc, --strand-collapse
                        Collapse counts across across Watson and Crick strands.
                        Only possible for CG methylation type. The strand is
                        recorded as '*' if this option is selected. (default:
                        False)
  --nfff, --no-failed-filter-file
                        Do not create the file listing the reads that failed to
                        pass to pass the filters and which filter it failed
                        (default: False)
  --gzip                gzip all output files. --gzip and --bzip2 are mutually
                        exclusive (default: False)
  --bzip2               bzip2 all output files. --gzip and --bzip2 are mutually
                        exclusive (default: False)

Construction of methylation loci m-tuples:
  --mt {CG,CHG,CHH,CNN}, --methylation-type {CG,CHG,CHH,CNN}
                        The methylation type. Multiple methylation types may be
                        analysed jointly by repeated use of this argument, e.g.,
                        --methylation-type CG --methylation-type CHG. The
                        default ('None') corresponds to CG (default: None)
  -m <int>              The size of the m-tuples, i.e., the 'm' in m-tuples
                        (default: 1)
  --ac, --all-combinations
                        Create all combinations of m-tuples, including non-
                        neighbouring m-tuples. WARNING: This will greatly
                        increase the memory usage, particularly for larger
                        values of -m and when analysing non-CG methylation
                        (default: False)

Filtering of reads:
  Applied before filtering of bases

  --id, --ignore-duplicates
                        Ignore reads that have been flagged as PCR duplicates
                        by, for example, Picard's MarkDuplicates function. More
                        specifically, ignore reads with the 0x400 bit in the
                        FLAG (default: False)
  --mmq <int>, --min-mapq <int>
                        Ignore reads with a mapping quality score (mapQ) less
                        than <int> (default: 0)
  --of {sequence_strict,sequence,XM_strict,XM,XM_ol,quality,Bismark}, --overlap-filter {sequence_strict,sequence,XM_strict,XM,XM_ol,quality,Bismark}
                        The type of check to be performed (listed roughly from
                        most-to-least stringent): Ignore the read-pair if the
                        sequence in the overlap differs between mates
                        (sequence_strict); Ignore the overlapping region if the
                        sequence in the overlap differs between mates
                        (sequence); Ignore the read-pair if the XM-tag in the
                        overlap differs (XM_strict); Ignore the overlapping
                        region if the XM-tag in the overlap differs between
                        mates (XM); Ignore any positions in the overlapping
                        region where the XM-tags differ between the mates
                        (XM_ol); Use the mate with the higher average quality
                        basecalls in the overlapping region (quality); Use the
                        first mate of each read-pair, i.e., the method used by
                        bismark_methylation_extractor with the --no_overlap flag
                        (Bismark) (default: XM_ol)
  --uip, --use-improper-pairs
                        Use the improper read-pairs, i.e. don't filter them.
                        More specifically, check the 0x2 FLAG bit of each read;
                        the exact definition of an improper read-pair depends on
                        the aligner and alignment parameters (default: False)

Filtering of bases:

NEWS.md
# Version 1.8.0

- Fixed parsing of reads containing insertions (closes [#104](https://github.com/PeteHaitch/methtuple/issues/104))
- Fixed handling of gzip and bzip output files and filenames
- Remove support for Python 2.7. It may still work, but I'm no longer going to try to maintain compatability with both Python 2 and 3.

# Version 1.7.0

- Fixed parsing of `--methylation-type` argument (closes [#102](https://github.com/PeteHaitch/methtuple/issues/102))

# Version 1.5.4

`methtuple` now requires pysam `v0.8.4` or greater.
`methtuple` now support Python 3.5.

## Bug fixes

- Fixed failing tests due to changes between pysam v0.8.3 and v0.8.4.

## Internal

- `methtuple` now requires pysam `v0.8.4` or greater

# Version 1.5.3

`methtuple` now requires pysam `v0.8.3` or greater.

## Bug fixes

- Fixed failing tests due to changes between pysam v0.8.2 and v0.8.3.

## Internal

- `methtuple` now requires pysam `v0.8.3` or greater

# Version 1.5.0

`methtuple` now requires `pysam v0.8.1` or greater.

## Bug fixes

* Fixed Python 3 compatibility issues due to changes in the `pysam` API (closes [#94](https://github.com/PeteHaitch/methtuple/issues/94)).

## Internal

Moved to the new `pysam` API. As a result, `methtuple` now requires `pysam v0.8.1` or greater.

# Version 1.4.0

This version of the software changes the name from `comethylation` to `methtuple` to better reflect what it actually does.

I have also switched to the MIT licence.

## New or improved features

* Added `--all-combinations` option to create all possible m-tuples including non-neighbouring ones. This will greatly increase the memory usage if combined with larger values of `--m` (e.g. > 2).

# Version 1.3.0

## New or improved features

* Support reads containing indels (closes [#25](https://github.com/PeteHaitch/comethylation/issues/25) and [#45](https://github.com/PeteHaitch/comethylation/issues/45)).
* Refined process for dealing with overlapping reads (closes [#15](https://github.com/PeteHaitch/comethylation/issues/15)).
* Improved documentation (closes [#82](https://github.com/PeteHaitch/comethylation/issues/82) and [#84](https://github.com/PeteHaitch/comethylation/issues/84)).
* Python 3.4 support.

## Minor improvements

* Added NEWS file.

## Bug fixes

* Provide default `--ir1p` and `--ir2p` (closes [#79](https://github.com/PeteHaitch/comethylation/issues/79)).
* Switched to *nix style line endings (\n), rather than the Windows line endings (\r\n), for the .tsv file (closes [#83](https://github.com/PeteHaitch/comethylation/issues/83)).
* Switched back to samtools instead of Picard in helper_scipts/run_comethylation.sh (closes [#78](https://github.com/PeteHaitch/comethylation/issues/78)).
* Updated examples to use Bismark v0.12.5.

## Internal

* Simplified creation of m-tuples.
* Added `get_positions` function to get the reference positions of all bases in a read (including `None` if no reference base, such as inserted bases or soft-clipped read-positions).
* Added `process_overlap` function to process overlapping paired-end reads using the given `--overlap-check`

## :(

* Running time of `v1.3` is 2-3x slower than `v1.2`. Partly because reads with indels are no longer skipped and partly because of code simplifications. Should be able to improve performance now that some internals have been simplified.

# Version 1.2.0

## New or improved features

* Python 3 support (not yet Python 3.4 compatible due to limitations of current version of Pysam).
* Include strand of m-tuple in output.

## Minor improvements
* Updated examples to use Bismark v0.12.2 (closes [#65](https://github.com/PeteHaitch/comethylation/issues/65)).
* Update `data/run_comethylation.sh` to use command line interface introduced in `v1.1`.
* Improved filtering of specific read-positions.

## Bug fixes
* Fixed bug that meant the sum of the overlap scores was incorrect for paired-end reads aligned to the OT-strand.


# Version 1.1.0

## New or improved features

* Complete re-design and simplification of command line interface (closes [#74](https://github.com/PeteHaitch/comethylation/issues/74))
* Reduced memory usage (addresses [#64](https://github.com/PeteHaitch/comethylation/issues/64)).
* Added support for gzip and bzip2 output files (closes [#69](https://github.com/PeteHaitch/comethylation/issues/69)).
* Added parallelisation via GNU parallel in `helper_scripts/run_comethylation.sh`(closes [#68](https://github.com/PeteHaitch/comethylation/issues/68)).
* Ignore specific read-positions rather than just start or end (closes [#71](https://github.com/PeteHaitch/comethylation/issues/71))

## Bug fixes
* `--methylationType CHG` requires `--strandSpecific` (closes [#72](https://github.com/PeteHaitch/comethylation/issues/72)).

## Internal

* Refactored `WithinFragmentComethylationMTuple` class as part of memory usage improvements.

# Version 1.0.0

Initial public release


Python/SAM2MS.py
#!/usr/bin/python
import re
import random
from numpy import array
import argparse
import sys
import csv
import pysam

## Extract comethylation signal for reads overlapping multiple CpGs
## SAM input file is the output of Bismark (no pre-processing required)
## CpG overlap is as determined by Bismark output, i.e. only positions with Z or z in the methylation string are considered CpGs
## This program is a re-write of SAM2MS_deprecated.py in order to utilise the pysam library. This version contains much additional functionality.


## This program is Copyright (C) 2012, Peter Hickey (hickey@wehi.edu.au)

## This program is free software: you can redistribute it and/or modify
## it under the terms of the GNU General Public License as published by
## the Free Software Foundation, either version 3 of the License, or
## (at your option) any later version.

## This program is distributed in the hope that it will be useful,
## but WITHOUT ANY WARRANTY; without even the implied warranty of
## MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
## GNU General Public License for more details.

## You should have received a copy of the GNU General Public License
## along with this program. If not, see <http://www.gnu.org/licenses/>.


# Order of fields in Bismark SAM file ###
  #####################################################################################
  ### (1) QNAME                                                                     ###
  ### (2) FLAG                                                                      ### 
  ### (3) RNAME                                                                     ###
  ### (4) POS                                                                       ###
  ### (5) MAPQ                                                                      ###
  ### (6) CIGAR                                                                     ###
  ### (7) RNEXT                                                                     ###
  ### (8) PNEXT                                                                     ###
  ### (9) TLEN                                                                      ###
  ### (10) SEQ                                                                      ###
  ### (11) QUAL                                                                     ###
  ### (12) NM-tag (edit distance to reference)                                      ###
  ### (13) XX-tag (base-by-base mismatches to the reference, not including indels)  ###
  ### (14) XM-tag (methylation call string)                                         ###
  ### (15) XR-tag (read conversion state for the alignment)                         ###
  ### (16) XG-tag (genome conversion state for the alignment)                       ###
  #####################################################################################

# Explanation of methylation string (XM) 
  #################################################################
  ### . for bases not involving cytosines                       ###
  ### X for methylated C in CHG context (was protected)         ###
  ### x for not methylated C in CHG context (was converted)     ###
  ### H for methylated C in CHH context (was protected)         ###
  ### h for not methylated C in CHH context (was converted)     ###
  ### Z for methylated C in CpG context (was protected)         ###
  ### z for not methylated C in CpG context (was converted)     ###
  #################################################################

### Stranded-ness in Bismark SAM files
# A read that aligns to the reverse strand has its read sequence reverse-complemented prior to being stored in the SEQ field of the SAM file (consistent with Bowtie)
# This means that the 5' end of the original read (aka the 'good quality end') is at the leftmost position in the SEQ field regardless of which strand the read aligned against.

### A note on duplicates
# True PCR duplicates may be missed by MarkDuplicates if the reads have undergone different amounts of trimming since the start and end co-ordinates of the duplicate reads no longer need coincide even if both reads have the same POS field in the SAM file.

## TODO: Fix ignore5 
## TODO: Extract strand information from XS tag, rather than relying on read orientation (which is the incorrect way to infer strand information).
## TODO: Add a count of the number of CpGs between the CpG-pair to the output for each CpG pair
## TODO: Add normalising factor to "position in read of (un)methylated CpGs" based on number of reads of length n
## TODO: Use "read.positions" to extract position of CpG in reference genome rather than my method of read.start + index. This method may(?) generalise better to reads containing indels, i.e. using the Bowtie2 pipeline
## TODO: Clarifiy how "read.positions" works if there is an insertion or deletion in the read; example read HWI-ST567_0245:2:65:16410:91305#GGCTAC in 46210.bam
## TODO: Discuss "random choice of CpGs in a read" with Terry - naive approach results in pairs with intra-pair distance > 40 (< 30%). Prove this analytically. Might choose outermost pair instead.
## TODO: Paired-end version. Be careful of orientation of read pairs and stranded-ness for ignore5, ignore3 options and parsing of XM tag
## TODO: Add flags to skip 'non-primary alignment' reads, 'failed reads', etc.

# Command line passer
parser = argparse.ArgumentParser(description='Extract the methylation calls for a CpG pair from reads that overlap multiple CpGs from a Bismark SAM/BAM file. The output file contains the positions of the CpG pair using 1-based co-ordinates on the forward strand of the cytosine in each CpG. If a read overlaps more than two CpGs there are several ways to construct the pairs (see the --pairChoice argument). The output of this file can be used for analysing comethylation along a read.')
parser.add_argument('SAM', metavar = 'SAM',
                  help='The path to the Bismark SAM/BAM file')
parser.add_argument('output', type=argparse.FileType('w'), nargs=1,
                   help='The output filename')
parser.add_argument('-5', '--ignore5', metavar = '<int>', type = int,
                  default=0,
                  help='Ignore <int> bases from 5\' (left) end of reads (default: 0). WARNING: Parameter value not sanity checked by program.')
parser.add_argument('-3', '--ignore3', metavar = '<int>', type = int,
                  default=0,
                  help='Ignore <int> bases from 3\' (right) end of reads (default: 0). WARNING: Parameter value not sanity checked by program.')
parser.add_argument('--maxReadLength', metavar = '<int>', type = int,
                  default=150,
                  help='Maximum read length in bp (default: 150)')
parser.add_argument('--pairChoice', metavar = '<string>',
                  default="random",
                  help='Method for constructing CpG pairs in reads that contain more than two CpGs: random, leftmost or outermost (default: random)')
parser.add_argument('--ignoreDuplicates', action='store_true',help='Ignore reads that have been flagged as PCR duplicates by Picard\'s MarkDuplicates function')

args = parser.parse_args()

# Create (possibly redundant) pointers to files from command line arguments
OUT = args.output[0]
maxReadLength = args.maxReadLength
pairChoice = args.pairChoice
ignore5 = args.ignore5
ignore3 = args.ignore3

# Variable initialisations
CpG_pattern = re.compile(r"[Zz]")
methylated_CpG_pattern = re.compile(r"[Z]")
unmethylated_CpG_pattern = re.compile(r"[z]")
read_counter = 0 # The number of reads processed by SAM2MS.py
duplicate_read_counter = 0 # The number of duplicate reads ignore by SAM2MS.py
CpGs_per_read = [0] * (maxReadLength/2 + 2) # The number of reads containing x CpGs (x = 0, 1, ...; the minimum number of CpGs is 0 and the maximum number of CpGs is half the read length + 1)
CpG_positions = [0] * maxReadLength # Count how often each position on the read was a CpG
methylated_CpG_positions = [0] * maxReadLength # Count how many times each position on the read was a methylated CpG
unmethylated_CpG_positions = [0] * maxReadLength # Count how many times each position on the read was an unmethylated CpG

# Function to write tab-separated outputfile
tabWriter = csv.writer(OUT, delimiter='\t', quotechar=' ', quoting=csv.QUOTE_MINIMAL)

# Opens SAM/BAM file
sam = pysam.Samfile(args.SAM) 

# Process each line of SAM file
for read in sam:
    # Skip duplicate reads if requested
    if args.ignoreDuplicates and read.is_duplicate:
        read_counter += 1
        duplicate_read_counter += 1
        continue
    else:
        read_counter += 1 # Increment the counter
        chrom = sam.getrname(read.tid) # 
        start = read.pos + 1 # read.pos is 0-based in accordance with BAM file specificiations and Python standards, regardless of whether the file is in BAM or SAM format. I convert this to 1-based positions.
        XM = [tag[1] for tag in read.tags if tag[0]=='XM'][0] # Tags are stored as a Python list of 2-tuples [("TAG_NAME", "TAG_VALUE"), ...]. This assumes there is one, and one only, XM tag.
        CpG_index = [m.start() for m in re.finditer(CpG_pattern, XM)] # Stores index of CpGs in read (C for Watson strand, G for Crick strand)
        methylated_CpG_index = [m.start() for m in re.finditer(methylated_CpG_pattern, XM)] # Stores index of methylated CpGs in read
        unmethylated_CpG_index = [m.start() for m in re.finditer(unmethylated_CpG_pattern, XM)] # Stores index of unmethylated CpGs in read
        n_CpGs = len(CpG_index) # The number of CpGs the read overlaps
        CpGs_per_read[n_CpGs] += 1 # Record how many CpGs were covered by the read
        # Increment the counters
        for i in CpG_index:
            CpG_positions[i] += 1
        for i in methylated_CpG_index:
            methylated_CpG_positions[i] += 1
        for i in unmethylated_CpG_index:
            unmethylated_CpG_positions[i] += 1
        # Ignore --ignore5 and --ignore3 bases from read
        CpG_index = [pos for pos in CpG_index if pos >= ignore5 and pos <= (len(XM) - ignore3)]
        n_CpGs = len(CpG_index) # Update the number of CpGs the read overlaps post-"ignore"
        # If there is more than one CpG in the read then we want to process that read
        if n_CpGs > 1:
            # Create a CpG pair - the choice is random, leftmost or outermost.
            # The outermost pair ensures the greatest intra-pair distance.
            if pairChoice=="random":
                CpG_pair = random.sample(CpG_index, 2)
                CpG_pair.sort() # Important to sort the result to ensure CpG_1 < CpG_2 by position
            elif pairChoice=="outermost":
                CpG_pair = [CpG_index[0], CpG_index[-1]]
            elif pairChoice=="leftmost":
                CpG_pair = [CpG_index[0], CpG_index[1]]
            else:
                sys.exit("Error: pairChoice must be one of random, leftmost or outermost. Please retry.")
            index = array(CpG_pair)
            # If a read maps to the reverse strand the Z/z characters in the XM string point to the G in the CpG - I want to point to the C in the CpG so I move the position coordinates 1bp to the left
            if(read.is_reverse):
                positions = start + index - 1 # positions are 1-based positions of the cytosines on the forward strand in each CpG
                strand = "-"
            else:
                positions = start + index # positions are 1-based positions of the cytosines on the forward strand in each CpG
                strand = "+"
            output=[chrom, positions[0], positions[1], XM[index[0]], XM[index[1]], strand]
            tabWriter.writerow(output)

# Close the file connections
sam.close()
OUT.close()

# Print summary statistics to standard output
print '%d reads processed' % read_counter
if args.ignoreDuplicates:
    print '%d duplicate reads ignored' % duplicate_read_counter
print '\nHistogram of CpGs per read'
sys.stdout.write('x')
sys.stdout.write('\t')
sys.stdout.write('Count\n')
print '------------------'
for index in range(len(CpGs_per_read)):
    print index, '\t', CpGs_per_read[index]
print '%d (%d%%) reads contained > 1 CpG (this includes ignored CpGs!)' % (sum(CpGs_per_read) - CpGs_per_read[0], 100 * (sum(CpGs_per_read) - CpGs_per_read[0])/read_counter)
print 'Position in read of methylated CpGs'
print methylated_CpG_positions
print 'Position in read of unmethylated CpGs'
print unmethylated_CpG_positions
print 'Position in read of CpGs'
print CpG_positions
